src/abstract_domains.rs

```diff
@@ -10,20 +10,100 @@ use constant_value::ConstantValue;
 /// A collection of abstract domains associated with a particular abstract value.
 /// The expression domain captures the precise expression that resulted in the abstract
 /// value. It can be used to derive any other kind of abstract domain on demand.
-#[derive(Serialize, Deserialize, Debug, Eq, PartialEq, Hash)]
+#[derive(Serialize, Deserialize, Clone, Debug, Eq, PartialEq, Hash)]
 pub struct AbstractDomains {
     pub expression_domain: ExpressionDomain,
     //todo: use cached getters to get other domains on demand
 }
 
-/// The things all abstract domain instances have to be able to do
+/// A collection of abstract domains that all represent the impossible abstract value.
+/// I.e. the corresponding set of possible concrete values is empty.
+pub const BOTTOM: AbstractDomains = AbstractDomains {
+    expression_domain: ExpressionDomain::Bottom,
+};
+
+/// A collection of abstract domains that all represent the universal abstract value.
+/// I.e. the corresponding set of possible concrete values includes every possible concrete values.
+pub const TOP: AbstractDomains = AbstractDomains {
+    expression_domain: ExpressionDomain::Top,
+};
+
+impl AbstractDomains {
+    /// True if all of the abstract domains in this collection correspond to an empty set of concrete values.
+    pub fn is_bottom(&self) -> bool {
+        self.expression_domain.is_bottom()
+    }
+
+    /// True if all of the abstract domains in this collection correspond to the set of all possible concrete values.
+    pub fn is_top(&self) -> bool {
+        self.expression_domain.is_bottom()
+    }
+
+    /// Joins all of the abstract domains in the two collections to form a single collection.
+    /// In a context where the join condition is known to be true, the result can be refined to be
+    /// just self, correspondingly if it is known to be false, the result can be refined to be just other.
+    pub fn join(
+        &self,
+        other: &AbstractDomains,
+        join_condition: &AbstractDomains,
+    ) -> AbstractDomains {
+        AbstractDomains {
+            expression_domain: self
+                .expression_domain
+                .join(&other.expression_domain, &join_condition),
+        }
+    }
+
+    /// Widen all of the abstract domains in the two collections to form a single collection.
+    pub fn widen(
+        &self,
+        other: &AbstractDomains,
+        join_condition: &AbstractDomains,
+    ) -> AbstractDomains {
+        AbstractDomains {
+            expression_domain: self
+                .expression_domain
+                .widen(&other.expression_domain, &join_condition),
+        }
+    }
+
+    /// True if for each abstract domains in the self collection, all of its concrete values are
+    /// elements of the corresponding set of the same domain in other.
+    pub fn subset(&self, other: &AbstractDomains) -> bool {
```

Comment:
I'd prefer a more abstract name like `less_than_equal`. You could imagine a case where your domain is underapproximate and `subset` is actually the superset operation... `less_than_equal` saves you from having to commit to a particular concretization and emphasizes the lattice structure of domains.

Reply:
I am not sure about this. I find it much easier to explain things in terms of a particular concretization and I do not currently anticipate a situation where join/widen will resulting in shrinking sets. Should things come to that, lots of things will have to change and it will be a suitable opportunity to change this name. For now, however, I find subset to be much more intuitive and explainable than less_than_or_equal.